File: sdks/nodejs/src/policy/decision.ts
export enum DecisionType {
  ALLOW = 'ALLOW',
  DENY = 'DENY',
  ALLOW_WITH_REDACTION = 'ALLOW_WITH_REDACTION',
  ALLOW_WITH_LIMITS = 'ALLOW_WITH_LIMITS',
  REQUIRE_APPROVAL = 'REQUIRE_APPROVAL',
  ALLOW_WITH_AUDIT_ONLY = 'ALLOW_WITH_AUDIT_ONLY',
  QUARANTINE = 'QUARANTINE',
}

export interface Decision {
  decision: DecisionType;
  reasonCode: string;
  riskScore: number;
  policyId?: string;
  redactionTargets?: string[];
  rateLimit?: { callsPerMinute?: number; tokensPerDay?: number };
  approvalRequiredFrom?: string;
  explanation?: string;
  latencyMs?: number;
}

export const isAllowed = (d: Decision): boolean =>
  [
    DecisionType.ALLOW,
    DecisionType.ALLOW_WITH_REDACTION,
    DecisionType.ALLOW_WITH_LIMITS,
    DecisionType.ALLOW_WITH_AUDIT_ONLY,
  ].includes(d.decision);

export const requiresRedaction = (d: Decision): boolean =>
  d.decision === DecisionType.ALLOW_WITH_REDACTION;

export class PolicyViolationError extends Error {
  constructor(public readonly decision: Decision) {
    super(
      `Policy violation: ${decision.reasonCode} (decision=${decision.decision})`
    );
    this.name = 'PolicyViolationError';
  }
}

File: sdks/go/policy/decision.go
// Package policy provides policy decision types and enforcement for CyberArmor.
package policy

import "fmt"

// DecisionType represents the outcome of a policy evaluation.
type DecisionType string

const (
	DecisionTypeAllow              DecisionType = "ALLOW"
	DecisionTypeDeny               DecisionType = "DENY"
	DecisionTypeAllowWithRedaction DecisionType = "ALLOW_WITH_REDACTION"
	DecisionTypeAllowWithLimits    DecisionType = "ALLOW_WITH_LIMITS"
	DecisionTypeRequireApproval    DecisionType = "REQUIRE_APPROVAL"
	DecisionTypeAllowWithAuditOnly DecisionType = "ALLOW_WITH_AUDIT_ONLY"
	DecisionTypeQuarantine         DecisionType = "QUARANTINE"
)

// Decision is the result of a policy evaluation.
type Decision struct {
	Type                 DecisionType `json:"decision"`
	ReasonCode           string       `json:"reason_code"`
	RiskScore            float64      `json:"risk_score"`
	PolicyID             string       `json:"policy_id,omitempty"`
	RedactionTargets     []string     `json:"redaction_targets,omitempty"`
	ApprovalRequiredFrom string       `json:"approval_required_from,omitempty"`
	Explanation          string       `json:"explanation,omitempty"`
	LatencyMs            int          `json:"latency_ms,omitempty"`
}

// IsAllowed returns true if the decision permits the action.
func (d Decision) IsAllowed() bool {
	switch d.Type {
	case DecisionTypeAllow, DecisionTypeAllowWithRedaction,
		DecisionTypeAllowWithLimits, DecisionTypeAllowWithAuditOnly:
		return true
	}
	return false
}

// RequiresRedaction returns true if response data must be redacted.
func (d Decision) RequiresRedaction() bool {
	return d.Type == DecisionTypeAllowWithRedaction
}

// PolicyViolationError is returned when a policy blocks an action.
type PolicyViolationError struct {
	Decision Decision
}

func (e *PolicyViolationError) Error() string {
	return fmt.Sprintf("CyberArmor policy violation: %s (%s)", e.Decision.ReasonCode, e.Decision.Type)
}
